### ISSUE TITLE
Login attempt without destroying the session

### ISSUE THREAD

**Issue by mzur** [getkirby/kirby#408]
Currently for every login attempt there is a [logout](https://github.com/getkirby/kirby/blob/d93de5fb29b3f1dc32f941cf46b6efdc27f3ba7a/core/user.php#L154) which destroys the session. This [breaks](https://github.com/mzur/kirby-uniform/issues/83) the Uniform login action when wrong login credentials were entered because Uniform stores a CSRF token in the session, too.

Maybe a less destructive way for clearing the session can be used. If I see that correctly only `kirby_auth_secret` and `kirby_auth_username` are [set for authentication](https://github.com/getkirby/kirby/blob/d93de5fb29b3f1dc32f941cf46b6efdc27f3ba7a/core/user.php#L164-L165) so these could be removed individually on logout without destroying the whole session, right?

**Comment by samnabi** [comment]
+1.

I'm experiencing this same issue with Shopkit: a user adds products to their cart, then wants to login in before paying, but the login process trashes the cart history.

**Comment by bastianallgeier** [comment]
We have worked on a solution for that, which you can find on the improved-login branch. We could need some help with testing it though, before we will merge it. Would be great to hear your thoughts.

**Comment by samnabi** [comment]
Thank you! I will hopefully have time to test this in the next couple weeks.

**Comment by mzur** [comment]
In the forum, Lukas pointed to the [OWASP article](https://www.owasp.org/index.php/Session_Management_Cheat_Sheet#Renew_the_Session_ID_After_Any_Privilege_Level_Change). The article only talks about renewing the session ID, not resetting the entire session ("session_start() & session_regenerate_id(true)"). If only the ID is updated, the session data will be kept and there is no need to manually preserve it. Did you have any other motivation for resetting the whole session?

Off topic: What is the difference between `logout` and `unauthorize` of `UserAbstract`?

**Comment by lukasbestle** [comment]
The issue was that the user was always logged out before logging in, even if no other user was logged in before. We still logout users on login, but only for that rare case. Otherwise only all Kirby-related session vars are deleted from the session.

> Off topic: What is the difference between logout and unauthorize of UserAbstract?

I can't tell for sure, but `unauthorize` seems to be used internally.

**Comment by bastianallgeier** [comment]
To call logout before login was just a lazy way of making sure that we are working with a fresh session tbh. User::unauthorize is really just internally and makes sure to remove all authentication tokens from the session.

**Comment by mzur** [comment]
Right, but my point is: according to OWASP there is no need to start a new session and manually copy the data. Something along the lines of this should be fine:

```php
public function login($password) {
  if(!password::match($password, $this->password)) return false; 

  $current = static::current();
  if ($current && !$this->is($current)) {
    // logout other user, destroy their session
    static::logout();
  }

  if (s::check()) {
    // regenerate session ID according to OWASP recommendations but keep session data
    session_regenerate_id(true);
  } else {
    s::start();
  }

  $key    = $this->generateKey();
  $secret = $this->generateSecret($key);
  s::set('kirby_auth_secret', $secret);
  s::set('kirby_auth_username', $this->username());
  cookie::set(
    s::$name . '_auth', 
    $key, 
    s::$cookie['lifetime'], 
    s::$cookie['path'], 
    s::$cookie['domain'], 
    s::$cookie['secure'], 
    s::$cookie['httponly']
  );
  return true;
}
```

The ` if (s::check()) else` block can probably be replaced by `s::regenerateId()`. That would be almost the same method as it was before except for the different logout behavior.

**Comment by lukasbestle** [comment]
@mzur @samnabi Have you found the time to test the new behavior yet?

**Comment by mzur** [comment]
I did. As I said in my previous post, I don't see the need to copy the session data, remove the `kirby_*` content and set the data again afterwards. For any active session `session_regenerate_id(true)` should be sufficient. The Kirby related content is overwritten later on anyway.

That said, your proposed code fixes the original issue for me.

**Comment by samnabi** [comment]
Just tested this with Shopkit, and the `improved-login` branch works perfectly for keeping the cart history intact.

I'm not familiar enough with sessions to do a code review on this, but from the end user perspective it looks great.

**Comment by lukasbestle** [comment]
Thanks for the feedback, guys. We will think about your suggestion.

**Comment by mzur** [comment]
Thanks, feel free to close this issue once everything is merged :+1:

**Comment by lukasbestle** [comment]
We have decided to keep the "clear Kirby session data" behavior to avoid potential security issues with the new feature. The feature is now merged and ready for 2.4.0. :)
